src/System.Net.Security/src/System/Net/NegotiateStreamPal.Windows.cs

```diff
@@ -0,0 +1,458 @@
+            IIdentity result = null;
+            string name = context.IsServer ? context.AssociatedName : context.Spn;
+            string protocol = context.ProtocolName;
+
+            if (context.IsServer)
+            {
+                SecurityContextTokenHandle token = null;
+                try
+                {
+                    SecurityStatusPal status;
+                    SafeDeleteContext securityContext = context.GetContext(out status);
+                    if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+                    {
+                        throw new Win32Exception((int)SecurityStatusAdapterPal.GetInteropFromSecurityStatusPal(status));
+                    }
+
+                    // This will return a client token when conducted authentication on server side.
+                    // This token can be used for impersonation. We use it to create a WindowsIdentity and hand it out to the server app.
+                    Interop.SecurityStatus winStatus = (Interop.SecurityStatus)SSPIWrapper.QuerySecurityContextToken(
+                        GlobalSSPI.SSPIAuth,
+                        securityContext,
+                        out token);
+                    if (winStatus != Interop.SecurityStatus.OK)
+                    {
+                        throw new Win32Exception((int)winStatus);
+                    }
+                    string authtype = context.ProtocolName;
+
+                    // TODO #5241:
+                    // The following call was also specifying WindowsAccountType.Normal, true.
+                    // WindowsIdentity.IsAuthenticated is no longer supported in CoreFX.
+                    result = new WindowsIdentity(token.DangerousGetHandle(), authtype);
+                    return result;
+                }
+                catch (SecurityException)
+                {
+                    // Ignore and construct generic Identity if failed due to security problem.
+                }
+                finally
+                {
+                    if (token != null)
+                    {
+                        token.Dispose();
+                    }
+                }
+            }
+
+            // On the client we don't have access to the remote side identity.
+            result = new GenericIdentity(name, protocol);
+            return result;
+        }
+
+        internal static string QueryContextAssociatedName(SafeDeleteContext securityContext)
+        {
+            return SSPIWrapper.QueryContextAttributes(GlobalSSPI.SSPIAuth, securityContext, Interop.SspiCli.ContextAttribute.Names) as string;
+        }
+
+        internal static string QueryContextAuthenticationPackage(SafeDeleteContext securityContext)
+        {
+            var negotiationInfoClass = SSPIWrapper.QueryContextAttributes(GlobalSSPI.SSPIAuth, securityContext, Interop.SspiCli.ContextAttribute.NegotiationInfo) as NegotiationInfoClass;
+            return negotiationInfoClass?.AuthenticationPackage;
```

Comment:
@stephentoub Last time I was reviewing, I remember that we were not allowing 6.0 language features. It would be nice to know it did and we can use the null-conditionals now.